File: shared/go/observability/metrics.go
package observability

import (
	"github.com/prometheus/client_golang/prometheus"
	"github.com/prometheus/client_golang/prometheus/promauto"
)

var telemetryExporterFailures = promauto.NewCounterVec(
	prometheus.CounterOpts{
		Name: "shared_telemetry_export_failures_total",
		Help: "Number of telemetry exporter initialization failures by exporter protocol.",
	},
	[]string{"service_name", "exporter"},
)

func recordExporterFailure(serviceName, exporter string) {
	if serviceName == "" {
		serviceName = "unknown"
	}
	telemetryExporterFailures.WithLabelValues(serviceName, exporter).Inc()
}

// TelemetryExporterFailures exposes the failure counter for integration tests and dashboards.
func TelemetryExporterFailures() *prometheus.CounterVec {
	return telemetryExporterFailures
}

File: shared/ts/src/observability/metrics.ts
import { Counter } from 'prom-client';

export const telemetryExporterFailures = new Counter({
  name: 'shared_telemetry_export_failures_total',
  help: 'Number of telemetry exporter initialization failures by protocol.',
  labelNames: ['service_name', 'exporter'] as const,
});

export function incrementTelemetryExporterFailure(serviceName: string, exporter: string) {
  telemetryExporterFailures.labels(serviceName || 'unknown', exporter).inc();
}

export function resetTelemetryMetrics(): void {
  telemetryExporterFailures.reset();
}

export async function telemetryExporterFailureCount(serviceName: string, exporter: string): Promise<number> {
  const metric = await telemetryExporterFailures.get();
  const target = metric?.values?.find(
    (value) =>
      value.labels.service_name === (serviceName || 'unknown') && value.labels.exporter === exporter,
  );
  return target?.value ?? 0;
}

